### ISSUE TITLE
Not hard fail on "Already Verified" hardhat-etherscan plugin

### ISSUE THREAD

**Issue by PatrickAlphaC** [NomicFoundation/hardhat#2287]
If you verify a contract programmatically and the contract is already verified, you'll get this horrible monstrosity of an output:

```
Successfully submitted source code for contract
contracts/SimpleStorage.sol:SimpleStorage at 0xA306f66B4275602be9ae0f7AfCf2f52bC7F55EBd
for verification on the block explorer. Waiting for verification result...

NomicLabsHardhatPluginError: The Etherscan API responded with a failure status.
The verification may still succeed but should be checked manually.
Reason: Already Verified
    at getVerificationStatus (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/@nomiclabs/hardhat-etherscan/src/etherscan/EtherscanService.ts:121:11)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async attemptVerification (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/@nomiclabs/hardhat-etherscan/src/index.ts:459:30)
    at async SimpleTaskDefinition.action (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/@nomiclabs/hardhat-etherscan/src/index.ts:741:48)
    at async Environment._runTaskDefinition (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/hardhat/src/internal/core/runtime-environment.ts:217:14)
    at async Environment.run (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/hardhat/src/internal/core/runtime-environment.ts:129:14)
    at async SimpleTaskDefinition.verifySubtask [as action] (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/@nomiclabs/hardhat-etherscan/src/index.ts:281:28)
    at async Environment._runTaskDefinition (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/hardhat/src/internal/core/runtime-environment.ts:217:14)
    at async Environment.run (/Users/patrick/code/hardhat-freecodecamp/hardhat-simple-storage-fcc/node_modules/hardhat/src/internal/core/runtime-environment.ts:129:14)
```

It would be great if there was a option to "ignore 'already verified'" flag on the verify command so it doesn't die.

**Comment by noahliechti** [comment]
I'm using this script to prevent the "monstrosity of an output". Nevertheless I'd like to the enhancement @PatrickAlphaC proposed.

```js
exports.verifyContract = async (contract, args) => {
  const { address } = contract;
  if (hre.network.config.chainId === 31337 || !hre.config.etherscan.apiKey) {
    return; // contract is deployed on local network or no apiKey is configured
  }
  console.log("Waiting 5 block confirmations...");
  await contract.deployTransaction.wait(5); // needed if verifyContract() is called immediately after deployment
  try {
    console.log("Verifying contract...");
    await hre.run("verify:verify", {
      address: address,
      constructorArguments: args,
    });
  } catch (err) {
    if (err.message.includes("Reason: Already Verified")) {
      console.log("Contract is already verified!");
    }
  }
};
```

**Comment by PatrickAlphaC** [comment]
I'm doing the same thing :) https://github.com/PatrickAlphaC/hardhat-fund-me-fcc/blob/62655bf37005cf3e8c633fdb3773c41aec348906/deploy/01-deploy-fund-me.js#L25

But still would be nice.

**Comment by cliffhall** [comment]
Yeah, I'm seeing this a lot on polygon.

**Comment by cliffhall** [comment]
This is still valid. When you archive the output of a script that does this, the output is just confusing. 

First of all, why is it "already verified" if I just deployed it and invoked the etherscan verify in the same script execution? JUST HOW? 

This is a failure inside the plugin and/or etherscan api. I've told it to verify something I deployed moments earlier. It **cannot** be "already verified". 

I suspect some kind of handshake / timeout / retry bungle.

**Comment by Sohailghafoor** [comment]
Error in plugin @nomiclabs/hardhat-etherscan: The Etherscan API responded with a failure status.
The verification may still succeed but should be checked manually.
Reason: Already Verified

**Comment by silasdavis** [comment]
@cliffhall yes it can, if the bytecode being verified is an exact match from the compiler output for the source you are submitting then it is verified once and forever from Etherscan's perspective. So all isntances of that bytecode will be associated with the first instance of the source code that produces a verification match. This is a feature not a bug.

It would be nice if this re-verifying was a no-op though, since you may not know if you or someone else has verified the same code in the past.

Catching the error and matching on a string works, but is ugly. A stable `err.code === 'ALREADY_VERIFIED'` would be nicer.

**Comment by cliffhall** [comment]
@silasdavis OK, yeah. That does make sense. I still feel strongly that this is something `hardhat-etherscan` should be handling internally. With perhaps a configuration parameter that lets me opt out of having this error converted to a simple success message.

If I was writing a script that *needed* to detect whether this contract had already been verified at a different address, that catching the error is the way to go. But I think that's probably the edge case.

**Comment by wei3erHase** [comment]
yo @cliffhall, what's happening is that etherscan is reading your contract as an exact copy of another that already was verified (probably already deployed it once?). that's the reason why if you deploy a new UniV3Pool, for example, the child comes already verified by etherscan.

but yeah, lots of possible ways in which etherscan could fail, it'd be nice not to stall the deployment scripts because of it

**Comment by jacob0123456** [comment]
I have got a same issue. Thank you.

**Comment by robercano** [comment]
Bumping this. Got same issue. Any chance that this could be fixed?

**Comment by bitcoinbrisbane** [comment]
> Bumping this. Got same issue. Any chance that this could be fixed?

Seems pretty straightforward.  The Etherscan api has methods to get the [abi or byte code ](https://docs.etherscan.io/api-endpoints/contracts) which returns a JSON object and 200.  There could be a flag to check that first.

Ill try make a PR this weekend.

```bash
curl https://api-goerli.etherscan.io/api?module=contract&action=getcontractcreation&contractaddresses=0x28F95Ee52F1b801467D0B46c984682177864Dfc8&apikey=<APIKEY>
```

```json
{
    "status": "1",
    "message": "OK",
    "result": [
        {
            "contractAddress": "0x28f95ee52f1b801467d0b46c984682177864dfc8",
            "contractCreator": "0x1ab4c6d9e25fc65c917afbefb4e963c400fb9814",
            "txHash": "0x6571688430afb7eca988c34a7a4ee032166a280012dbe8913018de2cd8b7a3e1"
        }
    ]
}
```

![image](https://user-images.githubusercontent.com/8411406/210119024-3050ac36-639c-494a-a880-92d57022355d.png)


### References
Network API endpoints https://docs.etherscan.io/v/goerli-etherscan/

**Comment by bitcoinbrisbane** [comment]
Oh, you can actually just check the verification status first 

```typescript
export async function getVerificationStatus(
  url: string,
  req: EtherscanCheckStatusRequest
): Promise<EtherscanResponse> {
  const parameters = new URLSearchParams({ ...req });
  const urlWithQuery = new URL(url);
  urlWithQuery.search = parameters.toString();

  const { request } = await import("undici");
  let response;
  try {
    response = await request(urlWithQuery, { method: "GET" });

    if (!(response.statusCode >= 200 && response.statusCode <= 299)) {
      // This could be always interpreted as JSON if there were any such guarantee in the Etherscan API.
      const responseText = await response.body.text();
      const message = `The HTTP server response is not ok. Status code: ${response.statusCode} Response text: ${responseText}`;

      throw new NomicLabsHardhatPluginError(pluginName, message);
    }
  } catch (error: any) {
    throw new NomicLabsHardhatPluginError(
      pluginName,
      `Failure during etherscan status polling. The verification may still succeed but
should be checked manually.
Endpoint URL: ${urlWithQuery.toString()}
Reason: ${error.message}`,
      error
    );
  }
```

**Comment by fvictorio** [comment]
That's very useful @bitcoinbrisbane, thanks. Feel free to send a PR.

**Comment by jakerockland** [comment]
> I still feel strongly that this is something `hardhat-etherscan` should be handling internally. With perhaps a configuration parameter that lets me opt out of having this error converted to a simple success message.

Very much agree – bump. 🙌

**Comment by jakerockland** [comment]
cc @ryley-o on this ticket :)

**Comment by OptimusOpus** [comment]
@bitcoinbrisbane encouraged me to have a go. I raised a PR [here](https://github.com/NomicFoundation/hardhat/pull/3609) if anyone has feedback

**Comment by fvictorio** [comment]
Fixed in #3675.
